### ISSUE TITLE
Recreate Look and Feel of Legacy Icons in svg

### ISSUE THREAD

**Issue by kevinhendricks** [Sigil-Ebook/Sigil#563]
Given Sigil 1.4.0 will support custom icon themes, it would be nice to have a legacy icon theme all in svg.

Work towards that end can be done here.

@BeckyDTP 
Attached is a open book svg (public domain according to the website) that could be unrotated and used as the basis for some of the missing icons.

Also attached are some open icon library svg files for some of the missing one.

**Comment by kevinhendricks** [comment]
[For_BeckyEbook.zip](https://github.com/Sigil-Ebook/Sigil/files/5327750/For_BeckyEbook.zip)
[legacy-progress-20201005.pdf](https://github.com/Sigil-Ebook/Sigil/files/5327751/legacy-progress-20201005.pdf)
[legacy-svg-draft-v01.zip](https://github.com/Sigil-Ebook/Sigil/files/5327752/legacy-svg-draft-v01.zip)

**Comment by kevinhendricks** [comment]
[some_other_legacy_svg_files.zip](https://github.com/Sigil-Ebook/Sigil/files/5327766/some_other_legacy_svg_files.zip)

**Comment by kevinhendricks** [comment]
I am happy to take a shot at filling in the gaps so to speak in your plan.  Just let me know.
An exact match is not required but if we could match the look and feel and most of the icons that would be great.

**Comment by kevinhendricks** [comment]
Okay, I have merged my stuff into BeckyEbooks, removed all of the svg width and height attributes, added the proper viewBox attributes where they were missing, saved all as plain svg (removing all of the inkscape stuff, and replaced a few git related ones with things that are a bit closer in style to the legacy icons and used an older insert-image icon that we used to use long ago at smaller sizes.
I repalce the change case icons with the chage case icons from main/ but with their fill colour changed to black.

So as far as I can tell, all that remains with embedded png files are these:

Create ToC
Edit Toc
Edit Metadata

And these 3 are based off the old BookView icon which does not exist in svg at all.  But I have found something close and I will see what I can do unless BeckyEbook or DiapDealer beats me to it.

See the attached legacy-svg-draft-v02.zip

**Comment by kevinhendricks** [comment]
[Uploading legacy-svg-draft-v02.zip…]()

**Comment by kevinhendricks** [comment]
Okay, I tried to recreate something close for generate-toc.svg, edit-toc.svg metadata-editor.svg from open-book.svg.  It is not perfect but it is close enough for government work (I hope).

Give my latest version a look-see and change or modify anything you want.
[legacy-svg-draft-v03.zip](https://github.com/Sigil-Ebook/Sigil/files/5329538/legacy-svg-draft-v03.zip)

**Comment by kevinhendricks** [comment]
Still lots of work could be done on refinements and closer matches but I think this set will get us going.

Feel free to change anything here.  Hope this helps.

KevinH

**Comment by BeckyDTP** [comment]
We were working in parallel a bit anyway, so I based the changes on draft-v3. I made the icons with the book (toc, metadata) practically identical, although I gave a different letter "i" in the center of the book. These are just details.

There are icons in the "changed" archive for various reasons: better compatibility with the original ones (help-browser), smoothing of lines (reload page), better readability (superscript, subscript), etc.

[legacy-svg-draft-v03-changed.zip](https://github.com/Sigil-Ebook/Sigil/files/5330145/legacy-svg-draft-v03-changed.zip)

In the "alternate" archive I have placed icons that are simply alternative – change case and spellcheck have an additional white border, which partially improves the look with a dark theme, but only partially. I absolutely don't think these "alternates" are better, just take a look.

[legacy-svg-draft-v03-alternate.zip](https://github.com/Sigil-Ebook/Sigil/files/5330143/legacy-svg-draft-v03-alternate.zip)

As you write – icons can be improved endlessly, but sometime you have to say STOP :)

Becky

**Comment by dougmassay** [comment]
Looking good! I just found out (to my immense pleasure) that while the compiled external binary resource files are OS specific (unfortunately), they don't appear to be architecture specific. At least on Windows anyway. I compiled the legacy icons with rcc.exe from the x86 version of Qt5.12.9, and they happily worked in a version of Sigil that was compiled with a 64-bit version of Qt5.12.9!

I don't have any easy way of testing the same on Linux, but t'm hoping we can reduce the number of files for a custom icon release to 3. One for each supported OS.

@kevinhendricks , can you get me a MacOS binary of rcc for Appveyor testing? The Appveyor MacOS images don't have Qt pre-installed, and I'm not going to install the whole shebang each time just to compile some icons with rcc. I'll just download the binary from somewhere and run it (if it's standalone on MacOS like it is on Windows).

**Comment by kevinhendricks** [comment]
I think that rcc should be platform agnostic.  BeckyEbook has posted a number of rcc files she made on Windows I believe and they load and work just fine on my macOS machine.

Are you sure the differences you are seeing between platforms matter?  Try grabbing a material.rcc file from one of the mac Travis builds and loading it as a custom user theme.  I bet it will work.
At least BeckyEbook's rcc files work just fine on my mac.

**Comment by kevinhendricks** [comment]
I can e-mail you a couple rcc files built on macOs tonight.

**Comment by kevinhendricks** [comment]
The Qt docs say these binary resource files are platform independent on their Qt website.  It should work on any platform.

**Comment by dougmassay** [comment]
An rcc I compiled with Linux wouldn't load for me on Windows, but maybe I did something wrong. I'll check again (or try it the other direction).

**Comment by kevinhendricks** [comment]
@BeckyDTP,
I will merge in your changed files into a v4 and most of your alt versions except for the git stuff as I wanted the git icons to be more like the open library svn checkout, diff and commit since that is where most of the legacy icons have come from.

I will post a v4 later tonight and we should be good to go I hope.

**Comment by kevinhendricks** [comment]
I checked the source code for rcc. There are version keys  stored inside the rcc file.  So Qt 5.12 built rcc files may not work on earlier Qt versions. And similarly Qt 5.15 built rcc files may not work on Qt 5.12 systems.

But rcc files built on earlier or equal Qt systems should work on later ones. 

It all depends when the version changes were made in rcc.

**Comment by dougmassay** [comment]
That's probably it. My Arch computer has 5.15.1. So maybe I should target something like Qt5.9.x. That way they should work for everything we try to support.

There's also a version option for rcc. Not sure what it's for, but I'll see if that can be used to any advantage.

Happy to be wrong! :)

**Comment by kevinhendricks** [comment]
Here are tha mac built main 3 rcc files.  Built with Qt 5.12.9 on MacOS 10.14.6

[mac_built_rcc_files.zip](https://github.com/Sigil-Ebook/Sigil/files/5330621/mac_built_rcc_files.zip)

**Comment by kevinhendricks** [comment]
And here are hopefully the final versions of the legacy icon theme.  We can then add these to our new Contributed Icon repo.
[Uploading legacy-svg-draft-v04.zip…]()

**Comment by kevinhendricks** [comment]
And for the record, here is the python program that BeckyDTP wrote to create different base colour versions of the material icon themes.

[ChangeIconColor-all.py.zip](https://github.com/Sigil-Ebook/Sigil/files/5330650/ChangeIconColor-all.py.zip)

**Comment by kevinhendricks** [comment]
Also, should we move wrapindicator.png out of all of the themes and store it in icon since it must stay a png file for the code to work and it is identical across all themes?  If so, I can make those changes tomorrow.

**Comment by kevinhendricks** [comment]
WHOPPS HIT THE DAMN CLOSE BUTTON BY MISTAKE AGAIN!

**Comment by kevinhendricks** [comment]
Twice - argh!

**Comment by dougmassay** [comment]
> Also, should we move wrapindicator.png out of all of the themes and store it in icon since it must stay a png file for the code to work and it is identical across all themes? If so, I can make those changes tomorrow.
I would think so, yes.

**Comment by BeckyDTP** [comment]
Draft v04 is unavailable. You probably sent a comment before the file was uploaded to github.
It was similar with the v02 draft.

**Comment by kevinhendricks** [comment]
Argh!  Will upload it again.

**Comment by kevinhendricks** [comment]
[legacy-svg-draft-v04.zip](https://github.com/Sigil-Ebook/Sigil/files/5335264/legacy-svg-draft-v04.zip)

**Comment by BeckyDTP** [comment]
One note:
Get pencil for **edit-toc** from **index-edit**.
The pencil is a shortened and simplified version of the one in **document-edit**.

**Comment by kevinhendricks** [comment]
I am away from my desktop machine with inkscape traveling today.  If you have a free moment, please go ahead and make that change and post a zip of the fixed icon here.  Otherwise, I will take a shot at it after the Thanksgiving holiday long weekend here in Canada.

Thanks.

**Comment by kevinhendricks** [comment]
FYI
- moved wrapindicator.png out of main, fluent, and material and into icon and modified the wrap code to find it.

Now the icon themes will be pure svg files which will be nice and removes the duplication.

**Comment by dougmassay** [comment]
When is that wrapindicator image visible? I see it's only referred to in wrapindicator.h, but I don't think I recall ever actually seeing the image when using Sigil.

**Comment by BeckyDTP** [comment]
I've already written about this to @kevinhendricks. Over the years of using Sigil, I have never seen this graphic.

**Comment by BeckyDTP** [comment]
I prepared scaled down versions of the four icons (with a book) to keep the form consistent for most (48px). 

_For clarification:_
Only a few are smaller (16px viewbox), but there is a good reason for this – such small icons only appear in the menu or submenu and are better rendered from small sizes. They are never displayed in full size, and even if they are, they are still very legible.
Such icons include, among others headers, icons for change case, help-browser…
The unique icons are the chevrons, which are based on the 24px size to display nicely.
[four-icons-with-book.zip](https://github.com/Sigil-Ebook/Sigil/files/5335553/four-icons-with-book.zip)

**Comment by kevinhendricks** [comment]
It is used in Find and Replace to indicate Search wrap. Open Sigil to a blank epub.  Type the following two lines each in there own p tag.
This is a line.
This is another line.
Click the cursor to the top.  Fire-up Find and Replace.
Find "line"
(it highlights "line")
Find next
(it highlights the second "line")

Then Find Next again and the wrap indicator will pop up briefly in the middle of the screen indicating the search has wrapped.  It is translucent.

Took me a while to find it!

**Comment by kevinhendricks** [comment]
Hi BeckyEbook,
I am confused.  The actual svg viewBox does nothing more than provide the internal coordinate system for the paths, transforms, etc.   The height and width determine the actual size of the drawing in pixels in the end.  Since floating point math is used, changing the viewBox to something that would make integer endpoints possible does not actually change things.  So if one viewBox uses 128x128 and another uses 16x16 given floating point precision (even with only 32 bit floats) makes them virtually the same.

Any issues you see are coming from the downsampling and conversion to pixmaps that is being done in Qt or at the platform level.

So you really should never need to change the viewBox unless you only want to use low precision floats or ints as path values but internally they are all double precision floating point values.

So exactly what differences are you seeing by changing the viewbox?

**Comment by BeckyDTP** [comment]
_Ad wrapindicator:_
hmmm… I don't see this indicator in Windows. ¯\\_(ツ)_/¯ Maybe I'm looking wrong. Can you catch it in the screenshot?

_Ad icons:_
I understand the files are vector and resizing makes no difference. After all, I've usually tried to unify the viewBox in all icons.
Your icons with the book were not even squares and hence my simplification, and by the way – unification.

**Comment by kevinhendricks** [comment]
Here is a screen shot that caught the wrap indicator on my macOS laptop.  Note, you must have the "wrap" option checked in Find and Replace to get the search to wrap around back to the top.
<img width="1792" alt="wrapindicator_in_action" src="https://user-images.githubusercontent.com/8493752/95256365-b3802f80-07f0-11eb-8f0d-1c6a42fcc9ec.png">

**Comment by BeckyDTP** [comment]
Thank you. Nice. But I can't see anyway. I wonder if other Windows users can see this?

**Comment by kevinhendricks** [comment]
Okay I added BeckyEbook's modified book icons and removed the wrapindicator.png and created legacy-svg-v05.zip
[legacy-svg-v05.zip](https://github.com/Sigil-Ebook/Sigil/files/5336589/legacy-svg-v05.zip)

**Comment by kevinhendricks** [comment]
Not sure why it can not be seen on Windows?  It is fleeting.  Took me 3 times to actually catch it in a screenshot.
Hmmm ... perhaps DiapDealer can test on his Windows and Linux to see if it ever appears.

**Comment by kevinhendricks** [comment]
Perhaps the time interval is too short.  You might want to edit Sigil/src/Misc/WrapIndicator.h
and modify the animation duration from 250 to something much longer to see if that helps make it visible on windows.

```
private slots:
    void runInternal() {
        QPropertyAnimation *anim = new QPropertyAnimation(this, "opacity", this);
        anim->setDuration(250);
        anim->setEndValue(0.);
        connect(anim, SIGNAL(finished()), this, SLOT(deleteLater()));
        anim->start(QAbstractAnimation::DeleteWhenStopped);
    }

```

**Comment by BeckyDTP** [comment]
I got it!
The key options are not only "Wrap" but also "Current file".

**Comment by BeckyDTP** [comment]
I'm surprised I never paid enough attention to it. Because even if I usually have "All files" selected, the indicator also appears when holding down the Ctrl key, ie when forcing the current file.
Now that I am aware, I immediately like this little thing in Sigil!

**Comment by dougmassay** [comment]
I haven't seen it on Linux, yet. Doesn't seem to matter whether it's set to Current File or All Files. Haven't checked on Windows.

This is sounding vaguely familiar now, though. I think I recall seeing it before.

**Comment by BeckyDTP** [comment]
I started Linux Arch and I can see an indicator. "Current file" (or holding down the Ctrl key) is necessary.

**Comment by dougmassay** [comment]
Weird. It's making an appearance now on my Linux machine. I swear I haven't changed a thing since I last tried it!

**Comment by kevinhendricks** [comment]
Any preference for the name of the repo to hold all of this on Sigil-Ebook:

https://github.com/Sigil-Ebook/IconThemes
https://github.com/Sigil-Ebook/ContributedIcons
https://github.com/Sigil-Ebook/ContributedThemes
https://github.com/Sigil-Ebook/Icons
https://github.com/Sigil-Ebook/Themes

Other?

**Comment by dougmassay** [comment]
I'm pretty name agnostic. :)

I'm more interested in what sort of layout/plan you're thinking of going with. With the notion of making github releases for multiple themes, my first instinct is to load each theme into its own branch, but I'm open to suggestions. I also don't envision new full releases for a particular icon theme if they get tweaked after their initial release. One release per theme that we can just update the binaries on if there's any future tweaking done. Thoughts?

**Comment by BeckyDTP** [comment]
The neutral name "IconThemes" is OK for me.

I also have questions:
- will there be three default sets in this new repository, "legacy" and any additional ones?
- will any changes/fixes/commits for single icons in default themes be implemented in the main release of new Sigil versions? Because in the future, you may need to add new icons to new functions and then I see various problems:
1. if someone has a "custom" theme, but it will not be compatible with the future version of Sigil, which will require new icons, what will happen then? Perhaps we will already schedule internal versioning of the icon theme files so that Sigil can "know" something is wrong and display to the user a sample message "this theme is for Sigil 1.4 and you need a Sigil 1.5 compatible theme".
2. perhaps a Readme.md file describing the "compatibility" of each individual theme could help.

- will there be "release" of binary versions of themes (as .rcc files ready for download)?

By the way:
I'd like to ask you to wait another day with the new repository. Tomorrow I will prepare changed versions of the three icons (chevron-down, chevron-up and widget-close) for "fluent" and "material" themes, which will be more compatible with other icons in the theme.
Since we did that in "legacy" theme, I would like to do it for "fluent" and "material". The icons will be gray (# 878787) to be neutral with the Sigil interface, but I want to refine their shape.

**Comment by kevinhendricks** [comment]
My thoughts on repo layout: IconThemes

- appveyor.xml (or travis)
- version.xml
- README.md (instructions on custom_icon_theme.rcc placement)
- ChangeLog.txt
- docs/create_rcc_on_mac.md
- docs/create_rcc_on_windows.md
- docs/create_rcc_on_linux.md
- src/ChangeIconColorAll.py
- src/fluent/fluent.qrc
- src/fluent/*.svg
- src/material/material.qrc
- src/material/*.svg
- src/main/main.qrc
- src/main/*.svg
- src/legacy/legacy.qrc
- src/legacy/*.svg
- CMakeLists.txt (to build rcc files for all variants including material colour variants.

Then we use AppVeyor and CMake (or Travis) to run the builds and create all of the rcc files.

These files are stored in this Repo's Releases.

If we decide to add a new icon, we copy it here and the push will signal rcc creation replacing the outdated one in releases.

Users who create their own full custom sets are on their own to keep them in sync with what Sigil needs.  They can always simply extract a new icon temporarily from whatever set they want to fill in.  Also, newicon sets will always be backwards compatible as the rcc files can have "extra" icons in it that older Sigil versions won't access.

So no versioning is needed.  In my tests with rcc missing icons, nothing bad happens, the icon is simply left blank.

Is that workable/doable?

All we have to remember is that if we create a new icon, we need 3 versions (main, fluent, and normal material).  We may want to add one to legacy as well but...

**Comment by kevinhendricks** [comment]
@BeckyDTP,
No problems in waiting.  DiapDealer and I have not even coordinated a date for the next release yet.  And I still need to update the plugin interface code to look for a "plugin.svg" file first and if none exists fallback to "plugin.png" code plus more testing.

I just wanted to get the new tepo up and going to see if creating releases of rcc files via Appveyor or Travis would even work.  We still have to worry about Qt and rcc versioning issues.

**Comment by kevinhendricks** [comment]
I am hoping that we will not need new icons much unless we plan to change the gui in a major way.

I am happy with Sigil now and I do not plan any big changes for the immediate future.  My Sigil dev goals and to-do list is now pretty much empty.  Mainly just refinements and bug fixes for the next couple of releases after the upcoming one unless inspiration for something new hits DiapDealer or I.

So hopefully these rcc files will hold is for the next year or so.

**Comment by jpmorgan58** [comment]
My 2¢ would be that custom themes should be compatible if they have all the elements of the standard themes, and that elements of a standard theme be used if a custom theme is missing them. Compatibility can be checked once when the theme is loaded. Version numbers make more sense for a database, where a standard interface can be used to access different versions, possibly through reformatting.

 - John

**Comment by kevinhendricks** [comment]
They actually map into the same resource tree node so loading both at the same time will not work.  This was done on purpose. We will keep the main themes up to date.  Creators of their own custom themes will be on their own to keep their themes in sync.  We have code to check for completeness if needed but this code really should berun by the icon theme developer not Sigil.

**Comment by dougmassay** [comment]
Ahhh... so your plan (@kevinhendricks ) is for a github Release to encompass all icon sets? As in each release would have all of the existing icon sets built and attached as assets to the release? If so, that would certainly be easier than doing a release for each icon set (which I admit, I was having a hard time getting my head around those logistics).

Using cmake might not even really be necessary when it boils down to it. There's nothing to configure, really. All we're doing is running rcc in a directory with a qrc file and a bunch of svgs. Windows Appveyor images have Qt5.9, which should be sufficient to ensure compatibility with all the systems we try to support. If not. I can install Qt5.9.5 on an Appveyor Ubuntu18.04 image and use that for the build. I can set it up so pushing a tag on the master branch will automatically compile the resources, create the github release and attach the assets. I do that with many of my plugin repos already. I just have it create the release as a draft. That way everything can be checked out, have release notes added before publishing.

**Comment by kevinhendricks** [comment]
@dougmassay,
Yes, exactly.  We can add or remove themes easily and each .rcc file would be an individual asset on the releases page.  Users can download only the .rcc files they want.

The nice thing is that github users can browse each icon theme to their heart's content before making a decision all from within a browser.

**Comment by dougmassay** [comment]
I think I'm getting a handle on if. ;) Just a couple more things:

Are the three "official" icon sets being included here for users to modify and/or base their new themes on, or will this be the definitive home of main, fluent, and material? In other words... will changes to main, fluent, and material flow TO Sigil or FROM Sigil?

I'm mainly wondering why would need to include rcc files for main, fluent, and material as part of a release, if they're already distributed with Sigil releases.

**Comment by kevinhendricks** [comment]
We need the material/ ones and BeckyDTP's  ChangeIconColorAll.py to generate the material derivatives (each with a different baser colour) in case people do not like that green or are colour blind to that color or just want another colour.

Since material and legacy have to be there, I just figured we would include main and fluent as well just in case users would like to mix and match icons in their own custom sets or are looking for inspiration.  The real home of main, fluent, and material (base green color) would be in the Sigil source code and this repo would just house local copies and any other user contributed sets down the road we like.

That said, we technically do not need fluent and main in this new repo at all if it creates a issue.  Nor do we need to create rcc files for fluent or main as those are built in the main Sigil build process.  We could of course but we do not need to.

So it is up to you.  Either way, the Sigil repo would be their true home.  We would copy and paste any new icons there into this new repo where appropriate.

**Comment by kevinhendricks** [comment]
FWIW, just pushed support for plugin.svg files for plugins to master, so we can start testing that.
The search order for plugin icons becomes (it stops as soon as it finds a valid icon)
    1. pluginsPath() + "/../plugins_prefs/" + name + "/plugin.svg";
    2. pluginsPath() + "/../plugins_prefs/" + name + "/plugin.png";
    3. pluginsPath() + "/" + name + "/plugin.svg";
    4. pluginsPath() + "/" + name + "/plugin.png";

The PluginDB which determines the icons is only really updated for existing plugins when Sigil is launched or the plugin is deleted and reinstalled.  So if you plan to test this feature, you will need to put the icon in place manually and then restart Sigil to see if your plugin.svg is properly detected.

**Comment by dougmassay** [comment]
I think we're square on the general plan! I see no reason the fluent and main being present should cause any problems. As I mentioned, we could always just choose not build/attach the rcc files for those Sigil-inherent icon sets in a new release of the custom icons.

We'll just have to be sure to decline any pull-requests to modify those three sets (since they're only going to be copies of the masters in Sigil).

**Comment by kevinhendricks** [comment]
Understood.  Do you want to take a shot at creating or should I?  I would need help on the appveyor parts.  We can update it with BeckyDTP's anytime before the release so no rush given we have no release date planned (yet).

**Comment by dougmassay** [comment]
Feel free to create it if you like. I'll fork it and experiment with Appveyor before turning everything loose on the official version.

**Comment by kevinhendricks** [comment]
I'll create it tomorrow and try to fully populate it with what we have so far.  I may use a CMake file to invoke the python script to create the material derivative sets since the only difference among them will be a fill colour hex code.

**Comment by kevinhendricks** [comment]
Took a shot at getting the basics up there: README.md, License info, and modified BeckyDTP's python code to work on source svg code and not on binary rcc files.  See src/ChangeMaterialThemeColor.py

It will work on the local material folder and change the svg files to be the desired color (one at a time) and copy it to a provided destination folder.

We can play around with it but the idea would be run this python program to create each color variation of material and then process them into rcc files along side legacy.

Let me know if this is enough for Appveyor to get started with.  If not, I can try to create a cmake file to do what we want.

**Comment by dougmassay** [comment]
I'll let you know. 

I'm going to add a gitignore file to avoid accidentally pushing .rcc files or other extraneous stuff to github.

**Comment by kevinhendricks** [comment]
Yes, my bad.  I forgot all about gitignore

**Comment by BeckyDTP** [comment]
Promised icons for material and fluent themes.
[chevrons_in_line_with_the_theme_icons.zip](https://github.com/Sigil-Ebook/Sigil/files/5348611/chevrons_in_line_with_the_theme_icons.zip)

I like the option of svg icons for plugins. Plugin developers can put both files (plugin.svg and plugin.png) if they want to be compatible with older Sigil versions. Sigil 1.4+ will use SVG and older versions will use PNG.

**Comment by kevinhendricks** [comment]
@BeckyDTP,
Merged chevrons_in_line_with_the_theme_icons.zip into master on IconThemes and Sigil.
And Thank you for the bug fix for the change color code.

**Comment by kevinhendricks** [comment]
I am going to close this as any further discussion can happen in the IconThemes repo by opening an issue.
